### PR TITLE
Support default factories taking validated data as an argument

### PR DESCRIPTION
**Pull request by Viicos**
## Change Summary

Closes https://github.com/pydantic/pydantic/issues/9789
Requires https://github.com/pydantic/pydantic-core/pull/1491

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by Viicos** [comment]
I haven't documented the `exclude_defaults` behavior as it seems pretty niche

File: pdm.lock [hunk 1]
```diff
@@ -1569,6 +1569,9 @@ files = [
 name = "pydantic-core"
 version = "2.25.0"
 requires_python = ">=3.8"
+git = "https://github.com/pydantic/pydantic-core"
+ref = "dependent-defaults"
+revision = "a3b4cf0cd5fd88e30da040b6aff12baeccecc702"
```
Comment (sydney-runkle) [inline pdm.lock:1574]:
```suggestion
```

File: pydantic/fields.py [hunk 2]
```diff
@@ -554,23 +555,36 @@ def deprecation_message(self) -> str | None:
...
 
         Returns:
             The default value, calling the default factory if requested or `None` if not set.
         """
         if self.default_factory is None:
             return _utils.smart_deepcopy(self.default)
         elif call_default_factory:
-            return self.default_factory()
+            if validated_data is None:
+                raise ValueError("'validated_data' must be provided if 'call_default_factory' is True.")
```
Comment (sydney-runkle) [inline pydantic/fields.py:582]: Is this true? Doesn't it only need to be not `None` when `_fields.takes_validated_data_argument(self.default_factory)` is True?

**Review by sydney-runkle — APPROVED** [review]
Nice work! One minor question.

Thanks for adding the section to the concepts docs, looks great.

File: pydantic/fields.py (hunk 2 above)
Comment (Viicos) [reply pydantic/fields.py:582]: Yes I can probably move the check below

Comment (sydney-runkle) [reply pydantic/fields.py:582]: Fine by me, feel free to merge then :).

**Comment by milani** [comment]
Thanks for adding this! Can PrivateAttr's `default_factory` work the same way?

**Pull request merged**